### PR TITLE
Fix issue Minetest crash when custom font path is not exist

### PR DESCRIPTION
**Pull request by srifqi**
We try to use default fallback for both mono and main font when custom font path is not exist. This way, if Minetest is not corrupted, we could avoid crash.

~Idea:~
- ~Put a dialog to tell the player. (?)~

(Not for now. Probably in another PR.)

### CODE REVIEW

**Comment by SmallJoker** [comment]
We already have these values in the `defaultsettings.cpp` file. Instead of copying these values, please extend `settings.cpp` to return the corresponding value from the `m_defaults` map.

**Comment by srifqi** [comment]
@SmallJoker how about now? I just add some methods to `Settings` class.

**Comment by nerzhul** [comment]
better !

File: src/fontengine.cpp [hunk 1]
```diff
@@ -346,19 +346,55 @@ void FontEngine::initFont(unsigned int basesize, FontMode mode)
...
+			if (font_path != fallback_font_path) {
+				// try original fallback font
+				errorstream << "FontEngine: failed to load custom fallback font: " << font_path << ", "
+						"trying to fall back to original fallback font" << std::endl;
+
+				font = gui::CGUITTFont::createTTFont(m_env,
+					fallback_font_path.c_str(), size, true, true,
+					font_shadow, font_shadow_alpha);
+
+				if (font != NULL) {
```
Comment (nerzhul) [inline src/fontengine.cpp]: if (font)

File: src/fontengine.cpp [hunk 2]
```diff
@@ -346,19 +346,55 @@ void FontEngine::initFont(unsigned int basesize, FontMode mode)
 			return;
 		}
 
-		// try fallback font
-		errorstream << "FontEngine: failed to load: " << font_path << ", trying to fall back "
-				"to fallback font" << std::endl;
+		if (font_config_prefix == "mono_") {
+			std::string mono_font_path = m_settings->getDefault("mono_font_path");
```
Comment (nerzhul) [inline src/fontengine.cpp]: const std::string &

File: src/fontengine.cpp [hunk 3]
```diff
@@ -346,19 +346,55 @@ void FontEngine::initFont(unsigned int basesize, FontMode mode)
...
-			font_path.c_str(), size, true, true, font_shadow,
-			font_shadow_alpha);
+			font = gui::CGUITTFont::createTTFont(m_env,
+				font_path.c_str(), size, true, true, font_shadow,
+				font_shadow_alpha);
 
-		if (font != NULL) {
-			m_font_cache[mode][basesize] = font;
-			return;
+			if (font != NULL) {
```
Comment (nerzhul) [inline src/fontengine.cpp]: if (font)

File: src/fontengine.cpp [hunk 4]
```diff
@@ -346,19 +346,55 @@ void FontEngine::initFont(unsigned int basesize, FontMode mode)
...
+				m_font_cache[mode][basesize] = font;
+				return;
+			}
+
+			std::string fallback_font_path = m_settings->getDefault("fallback_font_path");
```
Comment (nerzhul) [inline src/fontengine.cpp]: const std::string &

**Comment by SmallJoker** [comment]
:+1: from me once the minor issues above are fixed.

**Comment by srifqi** [comment]
Sorry for late response.

File: src/fontengine.cpp [hunk 5]
```diff
@@ -334,24 +334,60 @@ void FontEngine::initFont(unsigned int basesize, FontMode mode)
...
-		errorstream << "FontEngine: failed to load: " << font_path << ", trying to fall back "
-				"to fallback font" << std::endl;
+		if (font_config_prefix == "mono_") {
+			const std::string & mono_font_path = m_settings->getDefault("mono_font_path");
 
-		font_path = g_settings->get(font_config_prefix + "fallback_font_path");
+			if (font_path != mono_font_path) {
+				// try original mono font
+				errorstream << "FontEngine: failed to load custom mono font: " << font_path << ", "
+						"trying to fall back to original mono font" << std::endl;
```
Comment (SmallJoker) [inline src/fontengine.cpp]: Lines too long (>80 characters)

File: src/fontengine.cpp [hunk 6]
```diff
@@ -334,24 +334,60 @@ void FontEngine::initFont(unsigned int basesize, FontMode mode)
...
-			m_font_cache[mode][basesize] = font;
-			return;
+				if (font) {
+					m_font_cache[mode][basesize] = font;
+					return;
+				}
+			}
+		} else {
+			// try fallback font
+			errorstream << "FontEngine: failed to load: " << font_path << ", trying to fall back "
```
Comment (SmallJoker) [inline src/fontengine.cpp]: Too long as well

File: src/fontengine.cpp [hunk 7]
```diff
@@ -334,24 +334,60 @@ void FontEngine::initFont(unsigned int basesize, FontMode mode)
...
+				m_font_cache[mode][basesize] = font;
+				return;
+			}
+
+			const std::string & fallback_font_path = m_settings->getDefault("fallback_font_path");
+
+			if (font_path != fallback_font_path) {
+				// try original fallback font
+				errorstream << "FontEngine: failed to load custom fallback font: " << font_path << ", "
+						"trying to fall back to original fallback font" << std::endl;
```
Comment (SmallJoker) [inline src/fontengine.cpp]: Again too long

**Comment by SmallJoker** [comment]
Apart from the few comments above - the code looks good and :+1: once they're handled.

File: src/fontengine.cpp [hunk 8]
```diff
@@ -334,32 +334,70 @@ void FontEngine::initFont(unsigned int basesize, FontMode mode)
...
+			font = gui::CGUITTFont::createTTFont(m_env,
+				font_path.c_str(), size, true, true, font_shadow,
+				font_shadow_alpha);
+
+			if (font) {
+				m_font_cache[mode][basesize] = font;
+				return;
+			}
+
+			const std::string & fallback_font_path = m_settings->getDefault("fallback_font_path");
```
Comment (nerzhul) [inline src/fontengine.cpp]: no space after reference marker

Comment (srifqi) [inline src/fontengine.cpp]: Which part? `const std::string&` ?

Comment (red-001) [inline src/fontengine.cpp]: const std::string &fallback_font_path

**Comment by srifqi** [comment]
Sorry for late response, again.

**Comment by tobyplowy** [comment]
This wuld be a good feature for 0.5.0 :)

**Comment by sfan5** [comment]
9526c686997285700f500b975a9d971c3ca0a188

**Comment by srifqi** [comment]
Thanks for merge!

**Pull request closed without merging**